pkg/p2p/testbed/cluster_test.go

```diff
@@ -0,0 +1,153 @@
+// +build testbed
+
+package testbed
+
+import (
+	"context"
+	"encoding/hex"
+	"io"
+	"os/exec"
+	"strconv"
+	"sync"
+
+	"github.com/dusk-network/dusk-protobuf/autogen/go/rusk"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+type testNode struct {
+	kadcastPort int
+	grpcPort    int
+
+	exe *exec.Cmd
+
+	mu      sync.RWMutex
+	message []byte
+}
+
+func newNode(exePath string, bootstrapNodes []string, kadcastPort, grpcPort int) (*testNode, error) {
+	n := &testNode{
+		kadcastPort: kadcastPort,
+		grpcPort:    grpcPort,
+		message:     make([]byte, 0),
+	}
+
+	cmd := exec.Command(exePath,
+		"--ipc_method",
+		"tcp_ip",
+		"-p",
+		strconv.Itoa(grpcPort),
+		"--kadcast_public_address",
+		NetworkAddr+":"+strconv.Itoa(kadcastPort),
+		"--kadcast_bootstrap",
+		bootstrapNodes[0],
+		"--kadcast_bootstrap",
+		bootstrapNodes[1],
+	)
+
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	logrus.WithField("kadcastPort", kadcastPort).
+		WithField("grpc", grpcPort).
+		Info("Start a node")
+
+	n.exe = cmd
+
+	return n, nil
+}
+
+// GetMessage returns a copy of a received msg.
+func (t *testNode) GetMessage() []byte {
+	t.mu.RLock()
+	cpy := make([]byte, len(t.message))
+	copy(cpy, t.message)
+	t.mu.RUnlock()
+
+	return cpy
+}
+
+// Kill causes the Rusk Process to exit immediately.
+func (t *testNode) Kill() {
+	if t.exe != nil {
+		t.exe.Process.Kill()
+	}
+}
+
+// Broadcast sends broadcast grpc call from this node.
+func (t *testNode) Broadcast(ctx context.Context, payload []byte) error {
+	addr := NetworkAddr + ":" + strconv.Itoa(t.grpcPort)
+	conn, _ := createNetworkClient(ctx, addr)
+
+	m := &rusk.BroadcastMessage{
+		KadcastHeight: 128,
+		Message:       payload,
+	}
+	// broadcast message
+	if _, err := conn.Broadcast(ctx, m); err != nil {
+		log.WithError(err).Warn("failed to broadcast message")
+		return err
+	}
+
+	logrus.WithField("kadcast_port", t.kadcastPort).WithField("payload", hex.EncodeToString(payload)).
+		Info("Broadcast message")
+
+	return nil
+}
+
+func createNetworkClient(ctx context.Context, address string) (rusk.NetworkClient, *grpc.ClientConn) {
+	conn, err := grpc.DialContext(ctx, address, grpc.WithInsecure(), grpc.WithBlock())
+	if err != nil {
+		log.Panic(err)
+	}
+
+	return rusk.NewNetworkClient(conn), conn
+}
+
+// Listen creates a client connection and accepts stream messages.
+func (t *testNode) Listen(ctx context.Context) {
+	addr := NetworkAddr + ":" + strconv.Itoa(t.grpcPort)
+	conn, _ := createNetworkClient(ctx, addr)
+
+	// create stream handler
+	stream, err := conn.Listen(context.Background(), &rusk.Null{})
+	if err != nil {
+		log.WithError(err).Error("open stream error")
+		return
+	}
+
+	logrus.WithField("kadcast_port", t.kadcastPort).Info("Listening")
+
+	// accept  stream messages
+	for {
+		msg, err := stream.Recv()
+		if err != nil {
+			if err == io.EOF {
+				return
+			}
+
+			logrus.WithError(err).Error("recv error")
+			return
+		}
```

Comment:
Should we add a `select` block within the `for` loop to check for the `ctx.Done()` signal as well?

Reply:
Not needed. If you cancel the ctx that is passed to Listen, it will close the stream with `error="rpc error: code = Canceled desc = context canceled"`